### PR TITLE
[bug] use ports specified in eip-service.json

### PR DESCRIPTION
**Pull request by ivanalejandro0**
Resolves: #6541 - https://leap.se/code/issues/6541

### CODE REVIEW

**Comment by ivanalejandro0** [comment]
@elijh do we need to do any change on the `bitmask-root` file regarding ports?

**Comment by elijh** [comment]
no, i think bitmask-root should work fine:

````
ALLOWED_FLAGS = {
    "--remote": ["IP", "NUMBER", "PROTO"],
....
PARAM_FORMATS = {
    "NUMBER": lambda s: re.match("^\d+$", s),
````

**Comment by elijh** [comment]
it seems like it would be easy enough to also not hard code the protocol with this same change. it would be nice to be able to make tcp the default, since that will work for a lot more people.

**Comment by ivanalejandro0** [comment]
> it seems like it would be easy enough to also not hard code the protocol with this same change. it would be nice to be able to make tcp the default, since that will work for a lot more people.

You're right, I overlooked that.
Should I choose 'tcp' or 'udp' (in that order) depending which one is available like with the ports? Any other thing protocol specific that I need to have in mind?

**Comment by elijh** [comment]
well, that is a good question.

ideally, we would try udp first, then fall back to tcp. but, we don't have any code in bitmask to handle that yet, so maybe we should just stick to tcp if available.

however, it might just work if we issue two remote flags: "--remote ip port udp --remote ip port tcp" for every ip and port combination. openvpn is supposed to be smart enough to fail over on its own, although not as robustly as i would like.

i should do some testing to see what it does in these cases with multiple --remote flags using different protocols.

on the other hand, tcp will work more often because of MTU issues...

**Comment by ivanalejandro0** [comment]
@elijh any luck with the test for this tcp/udp thing?
Otherwise I can merge this fix for the ports and create a new issue to remove the hardcoded protocol.

**Comment by micah** [comment]
Ivan Alejandro <notifications@github.com> writes:

> @elijh any luck with the test for this tcp/udp thing?
> Otherwise I can merge this fix for the ports and create a new issue to remove the hardcoded protocol.

+1 for separating out issues and resolving this so that the other issue
can be dealt with independently and doesn't block this

**Comment by ivanalejandro0** [comment]
I've created https://leap.se/code/issues/6816 to track the tcp/upd issue.

File: src/leap/bitmask/services/eip/eipconfig.py [hunk 1]
```diff
@@ -302,6 +302,21 @@ def get_gateway_ip(self, index=0):
             logger.error("Invalid ip address in config: %s" % (ip_addr_str,))
             return None
 
+    def get_gateway_ports(self, index=0):
+        """
+        Return the ports of the gateway.
+
```
Comment (kalikaneko) [inline src/leap/bitmask/services/eip/eipconfig.py]: document index param

File: src/leap/bitmask/services/eip/vpnlauncher.py [hunk 2]
```diff
@@ -119,24 +119,40 @@ def get_gateways(kls, eipconfig, providerconfig):
         :param providerconfig: provider specific configuration
         :type providerconfig: ProviderConfig
 
-        :rtype: list
+        :rtype: list of tuple(gateway, port)
```
Comment (kalikaneko) [inline src/leap/bitmask/services/eip/vpnlauncher.py]: I would document this in returns, type is just list..

File: src/leap/bitmask/services/eip/vpnlauncher.py [hunk 3]
```diff
@@ -119,24 +119,40 @@ def get_gateways(kls, eipconfig, providerconfig):
...
             raise VPNLauncherException('No gateway was found!')
 
-        logger.debug("Using gateways ips: {0}".format(', '.join(gateways)))
+        for idx, gw in enumerate(gws):
+            ports = eipconfig.get_gateway_ports(idx)
+
+            the_port = "1194"  # default port
+
+            # pick the port preferring this order:
+            for port in ("443", "80", "53", "1193"):
```
Comment (kalikaneko) [inline src/leap/bitmask/services/eip/vpnlauncher.py]: extract tuple to a more meaningful named constant, at module level?

File: src/leap/bitmask/services/eip/vpnprocess.py [hunk 4]
```diff
@@ -961,8 +961,14 @@ def getGateways(self):
...
         """
-        gateways = self._launcher.get_gateways(
+        gws = self._launcher.get_gateways(
             self._eipconfig, self._providerconfig)
+
+        # filter out ports since we don't need that info
+        gateways = []
+        for gw, port in gws:
+            gateways.append(gw)
+
```
Comment (kalikaneko) [inline src/leap/bitmask/services/eip/vpnprocess.py]: I think it's more elegant to do that in the form of a list comprenhension:

```
[k for k, v in [("foo", 1), ("bar", 2)]]
```

**Comment by kalikaneko** [comment]
@ivanalejandro0 mind to add a little explanation of the change in the body of the commit msg?

File: src/leap/bitmask/services/eip/eipconfig.py (hunk 1 above)
Comment (ivanalejandro0) [inline src/leap/bitmask/services/eip/eipconfig.py]: added

File: src/leap/bitmask/services/eip/vpnlauncher.py (hunk 2 above)
Comment (ivanalejandro0) [inline src/leap/bitmask/services/eip/vpnlauncher.py]: fixed

File: src/leap/bitmask/services/eip/vpnlauncher.py (hunk 3 above)
Comment (ivanalejandro0) [inline src/leap/bitmask/services/eip/vpnlauncher.py]: fixed

File: src/leap/bitmask/services/eip/vpnprocess.py (hunk 4 above)
Comment (ivanalejandro0) [inline src/leap/bitmask/services/eip/vpnprocess.py]: alright, changed.

**Comment by meskio** [comment]
Revased and merged.

**Pull request closed without merging**